### PR TITLE
#17 - implement Wallet.File.ledger()

### PR DESCRIPTION
**Pull request by t-izbassar**
Fix for #17. Implemented ledger() method. Updated relevant test case.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#30](https://github.com/zold-io/java-api/issues/30) is now in scope, role is `REV`

**Comment by 0crat** [comment]
Job [#30](https://github.com/zold-io/java-api/issues/30) is already in scope

**Comment by 0crat** [comment]
This pull request [#30](https://github.com/zold-io/java-api/issues/30) is assigned to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), here is [why](https://www.0crat.com/footprint/CAZUREFND/1921857632); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @llorllale[/z](https://www.0crat.com/u/llorllale) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/java/io/zold/api/Wallet.java [hunk 1]
```diff
@@ -140,13 +144,24 @@ public Wallet merge(final Wallet other) {
...
-        //  the transactions in this copy of the wallet. Also add a unit test
-        //  to replace WalletTest.ledgerIsNotYetImplemented().
         @Override
-        public Iterable<Transaction> ledger() {
-            throw new UnsupportedOperationException(
-                "ledger() not yet supported"
+        public Iterable<Transaction> ledger() throws IOException {
+            return new Mapped<>(
+                txt -> new RtTransaction(
+                    txt.asString()
```
Comment (carlosmiranda) [inline src/main/java/io/zold/api/Wallet.java]: @t-izbassar we can move this (as well as the next line) to the previous line.

File: src/main/java/io/zold/api/Wallet.java [hunk 2]
```diff
@@ -69,11 +71,13 @@
...
      * This wallet's ledger.
      * @return This wallet's ledger
+     * @throws IOException If an IO error occurs
      */
-    Iterable<Transaction> ledger();
+    Iterable<Transaction> ledger() throws IOException;
 
     /**
      * Default File implementation.
+     * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
```
Comment (carlosmiranda) [inline src/main/java/io/zold/api/Wallet.java]: @t-izbassar `2 lines` is fine here (up to class declaration)

File: src/test/java/io/zold/api/WalletTest.java [hunk 3]
```diff
@@ -85,8 +93,11 @@ private Path wallet(final long id) throws IOException {
                     "zold",
                     "1",
                     hex,
-                    // @checkstyle LineLength (1 line)
-                    "MIGeMA0GCSqGSIb3DQEBAQUAA4GMADCBiAKBgGZCr/9hBChqsChd4sRAIpKNRinjhSW+J+S7PU5malVMiRHVoKjeooLDpWpij0A6vkzOvjrMldAZT0Fzgp0cJ15TOVwiQanQ5WuQDgRkLoxrdh/qyBApoDvk4OUEozOQPNwfpZOFfaUALPsPnv9995TlY9WcdSKW5dj041p1tJmlAgMBAAE="
+                    // @checkstyle LineLength (5 lines)
+                    "MIGeMA0GCSqGSIb3DQEBAQUAA4GMADCBiAKBgGZCr/9hBChqsChd4sRAIpKNRinjhSW+J+S7PU5malVMiRHVoKjeooLDpWpij0A6vkzOvjrMldAZT0Fzgp0cJ15TOVwiQanQ5WuQDgRkLoxrdh/qyBApoDvk4OUEozOQPNwfpZOFfaUALPsPnv9995TlY9WcdSKW5dj041p1tJmlAgMBAAE=",
```
Comment (carlosmiranda) [inline src/test/java/io/zold/api/WalletTest.java]: @t-izbassar this file is eventually going to become very cumbersome. Let's move it to a static file in `src/test/resources`. That way we can add additional wallet files for different tests.

File: src/test/java/io/zold/api/WalletTest.java [hunk 4]
```diff
@@ -69,9 +69,17 @@ public void mergeIsNotYetImplemented() throws IOException {
...
-    public void ledgerIsNotYetImplemented() throws IOException {
-        new Wallet.File(this.folder.newFile().toPath()).ledger();
+    @Test
+    @SuppressWarnings("PMD.AvoidDuplicateLiterals")
+    public void walletShouldBeAbleToReturnLedger() throws Exception {
+        MatcherAssert.assertThat(
+            new Wallet.File(this.wallet(12345L)).ledger(),
+            Matchers.containsInAnyOrder(
+                // @checkstyle LineLengthCheck (2 lines)
+                new RtTransaction("003a;2017-07-19T21:24:51Z;ffffffff9c0ccccd;Ui0wpLu7;98bb82c81735c4ee;For services;SKMPrVj..."),
```
Comment (carlosmiranda) [inline src/test/java/io/zold/api/WalletTest.java]: @t-izbassar sufficient to just check number of returned transactions here. These are dummy transactions that are invalidly formatted anyhow.

**Comment by t-izbassar** [comment]
@carlosmiranda updated.

File: src/main/java/io/zold/api/Wallet.java [hunk 5]
```diff
@@ -69,11 +71,13 @@
     /**
      * This wallet's ledger.
      * @return This wallet's ledger
+     * @throws IOException If an IO error occurs
      */
-    Iterable<Transaction> ledger();
+    Iterable<Transaction> ledger() throws IOException;
```
Comment (llorllale) [inline src/main/java/io/zold/api/Wallet.java]: @t-izbassar I don't think we need to declare `IOException` here

**Comment by t-izbassar** [comment]
@llorllale merge?

**Comment by llorllale** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
Job `gh:zold-io/java-api#30` is not assigned, can't get performer

**Comment by 0crat** [comment]
The job [#30](https://github.com/zold-io/java-api/issues/30) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @llorllale[/z](https://www.0crat.com/u/llorllale)

**Comment by ypshenychka** [comment]
@carlosmiranda According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Two out of four found issues are cosmetic. 
Please confirm that you'll try to find at least three major problems while future reviews.

**Comment by carlosmiranda** [comment]
@ypshenychka confirmed

**Comment by ypshenychka** [comment]
@carlosmiranda thanks

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 point(s) just awarded to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Comment by g4s8** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/zold-io/java-api/issues/30#issuecomment-404813584))

@g4s8 This is what I know about this job in [CAZUREFND](https://www.0crat.com/p/CAZUREFND), as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#30](https://github.com/zold-io/java-api/issues/30) is not in scope
  * Job [footprint](https://www.0crat.com/footprint/CAZUREFND?q=%7Bjob%3A%27gh%3Azold-io%2Fjava-api%2330%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Pull request merged**